### PR TITLE
Align unknown type handling with upstream

### PR DESCRIPTION
**Pull request by dgkimura**
This is another approach to solve same as PR #9655 where instead of backporting from postgres 10 we rip out the cleverness.

In Postgres / Greenplum, a literal (e.g. NULL or 'text') in a query with
no obvious type gets a pseudo-type "unknown". A column that refers to
such literals will also get the "unknown" type.

We used to have logic to infer type of unknown literals in a subquery
based on the context. For example, the following query returns 124 as
foo is inferred to be an integer:

```sql
select foo + 123 from (select '1' as foo) a;
```

As another example, Greenplum infers the type of NULL as int in the
following "INSERT ... SELECT" with unknown-typed columns from a nested
subquery:

```sql
-- CREATE TABLE foo (a int, b text, c int);

INSERT INTO foo
SELECT * FROM (
  '1',
  'bbb',
  NULL
) bar(x,y,z);
```

While this user experience is nice, this cleverness is flawed.
Specifically, given an operator expression (let's say it's an addition)
where one operand has "unknown" type, and the other operand has type T
(let's say int), it's questionable that we can infer the unknown must be
a T (an int). In fact, the number of operators that fit the pattern of
"+(T, ?)" could be either ambiguous, or nonexistent.

Of course, there are scenarios where types are clearer: the SELECT list
that feeds into an INSERT or UPDATE has to be compatible (read:
cast-ible) with the target columns. Similarly, the SELECT list that goes
into a set operation (e.g. UNION) has to be compatible with
corresponding columns from other sub-queries. Upstream Postgres already
handles a large part of these cases.

This commit removes the Greenplum-specific cleverness. See the updated
tests for the behavior changes. Because this mostly changes
parse-analysis, expressions that are already stored in the catalog are
_not_ impacted (read: we will be able to restore the output of pg_dump).

Not that it will give us feature parity, but in Postgres 10, we will
coerce remaining unknown typed literals to text. See Postgres commit
1e7c4bb00497.

Co-authored-by: Jesse Zhang <sbjesse@gmail.com>

### CODE REVIEW

**Comment by ashwinstar** [comment]
Just wish to make sure we are not getting master branch it state by this PR where it has dependency for PostgreSQL 12 (or 10) merge to complete and be merged?

**Comment by d** [comment]
> Just wish to make sure we are not getting master branch it state by this PR where it has dependency for PostgreSQL 12 (or 10) merge to complete and be merged?

For sure. We've updated the commit message (and the PR message) to include justification for the removal

**Comment by hlinnaka** [comment]
This is ready to be pushed, right?

File: src/test/regress/sql/strings.sql [hunk 1]
```diff
@@ -1,3 +1,7 @@
+-- start_matchsubs
```
Comment (hlinnaka) [inline src/test/regress/sql/strings.sql]: Let's add a GPDB_10_MERGE_FIXME comment here, along the lines of
```
-- GPDB_10_MERGE_FIXME: this shouldn't be needed after merge of upstream commit 1e7c4bb004. Remove when we get there.
```

Comment (dgkimura) [reply src/test/regress/sql/strings.sql]: Sounds good. Will do.

**Comment by dgkimura** [comment]
On Tue, Mar 10, 2020 at 5:41 AM Heikki Linnakangas wrote:
>
> This is ready to be pushed, right?

Yup, I'll squash and push today. Thanks for reviewing!

**Pull request merged**